### PR TITLE
Added 1.18 support

### PR DESCRIPTION
**Pull request by md5sha256**
Update poms to use spigot 1.18
Update the dough-skins:
- add a new PlayerHeadAdapter18 for 1.18 mappings

### CODE REVIEW

**Comment by md5sha256** [comment]
> You need to add 1.18 here

fixed

File: dough-skins/src/main/java/io/github/bakedlibs/dough/skins/nms/PlayerHeadAdapter.java [hunk 1]
```diff
@@ -26,9 +26,12 @@
...
 
-            if (version.isAtLeast(1, 17)) {
-                // 1.17+ mappings
-                return new PlayerHeadAdapterAfter17();
+            if (version.getMajorVersion() == 1 && version.getMinorVersion() == 17) {
+                // 1.17 mappings
+                return new PlayerHeadAdapter17();
+            } else if (version.isAtLeast(1, 18)) {
+                // 1.18 mappings
+                return new PlayerHeadAdapter18();
```
Comment (WalshyDev) [inline dough-skins/src/main/java/io/github/bakedlibs/dough/skins/nms/PlayerHeadAdapter.java:34]: How come you're doing this over 
```suggestion
            if (version.isAtLeast(1, 18)) {
                // 1.18 mappings
                return new PlayerHeadAdapter17();
            } else if (version.isAtLeast(1, 17)) {
                // 1.17 mappings
                return new PlayerHeadAdapter18();
```

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: dough-skins/src/main/java/io/github/bakedlibs/dough/skins/nms/PlayerHeadAdapter.java [hunk 2]
```diff
@@ -26,9 +26,12 @@
...
 
-            if (version.isAtLeast(1, 17)) {
-                // 1.17+ mappings
-                return new PlayerHeadAdapterAfter17();
+            if (version.isAtLeast(1, 18)) {
+                // 1.18 mappings
+                return new PlayerHeadAdapter17();
+            } else if (version.isAtLeast(1, 17)) {
+                // 1.17 mappings
+                return new PlayerHeadAdapter18();
```
Comment (TheBusyBiscuit) [inline dough-skins/src/main/java/io/github/bakedlibs/dough/skins/nms/PlayerHeadAdapter.java:34]: These are swapped

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by WalshyDev** [comment]
@md5sha256 What else is there left to do here?

**Comment by qwertyuioplkjhgfd** [comment]
> @md5sha256 What else is there left to do here?

create ``items.nms.ItemNameAdapterAfter18`` since it's changed from 17

**Comment by WalshyDev** [comment]
does that actually have issues... from my testing pre-1.18 release it did not

**Comment by qwertyuioplkjhgfd** [comment]
i've gotten this error before

```
[23:22:33] [Server thread/WARN]: java.lang.NullPointerException: Cannot invoke "java.lang.reflect.Method.invoke(Object, Object[])" because "this.getName" is null
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.nms.ItemNameAdapterAfter17.getName(ItemNameAdapterAfter17.java:31)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.ItemUtils.getItemName(ItemUtils.java:46)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem.getItemName(SlimefunItem.java:893)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.isSearchFilterApplicable(SurvivalSlimefunGuide.java:392)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.openSearch(SurvivalSlimefunGuide.java:361)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.core.guide.SlimefunGuide.openSearch(SlimefunGuide.java:92)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.lambda$createHeader$16(SurvivalSlimefunGuide.java:602)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInput$1.onChat(ChatInput.java:58)
[23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInputListener.lambda$checkInput$1(ChatInputListener.java:79)
[23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftTask.run(CraftTask.java:101)
[23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:483)
[23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1547)
[23:22:33] [Server thread/WARN]: 	at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:480)
[23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1470)
[23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1264)
[23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317)
[23:22:33] [Server thread/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by qwertyuioplkjhgfd** [comment]
looks like that's because it's no longer at ``nms.world.item.ItemStack#getName``
should also check that toString is at ``nms.network.chat.IChatBaseComponent#getString``

**Comment by md5sha256** [comment]
I haven't had time to test yet so I have no idea whats left :p That being said, I'll have a look at that issue later tomorrow or wed

**Comment by md5sha256** [comment]
> i've gotten this error before
> 
> ```
> [23:22:33] [Server thread/WARN]: java.lang.NullPointerException: Cannot invoke "java.lang.reflect.Method.invoke(Object, Object[])" because "this.getName" is null
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.nms.ItemNameAdapterAfter17.getName(ItemNameAdapterAfter17.java:31)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.ItemUtils.getItemName(ItemUtils.java:46)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem.getItemName(SlimefunItem.java:893)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.isSearchFilterApplicable(SurvivalSlimefunGuide.java:392)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.openSearch(SurvivalSlimefunGuide.java:361)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.core.guide.SlimefunGuide.openSearch(SlimefunGuide.java:92)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.lambda$createHeader$16(SurvivalSlimefunGuide.java:602)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInput$1.onChat(ChatInput.java:58)
> [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInputListener.lambda$checkInput$1(ChatInputListener.java:79)
> [23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftTask.run(CraftTask.java:101)
> [23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:483)
> [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1547)
> [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:480)
> [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1470)
> [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1264)
> [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317)
> [23:22:33] [Server thread/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
> ```

wait... this looks like its using an old version of dough and not this branch.

**Comment by qwertyuioplkjhgfd** [comment]
> > i've gotten this error before
> > ```
> > [23:22:33] [Server thread/WARN]: java.lang.NullPointerException: Cannot invoke "java.lang.reflect.Method.invoke(Object, Object[])" because "this.getName" is null
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.nms.ItemNameAdapterAfter17.getName(ItemNameAdapterAfter17.java:31)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.items.ItemUtils.getItemName(ItemUtils.java:46)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem.getItemName(SlimefunItem.java:893)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.isSearchFilterApplicable(SurvivalSlimefunGuide.java:392)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.openSearch(SurvivalSlimefunGuide.java:361)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.core.guide.SlimefunGuide.openSearch(SlimefunGuide.java:92)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.implementation.guide.SurvivalSlimefunGuide.lambda$createHeader$16(SurvivalSlimefunGuide.java:602)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInput$1.onChat(ChatInput.java:58)
> > [23:22:33] [Server thread/WARN]: 	at Slimefun v4.9-UNOFFICIAL.jar//io.github.thebusybiscuit.slimefun4.libraries.dough.chat.ChatInputListener.lambda$checkInput$1(ChatInputListener.java:79)
> > [23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftTask.run(CraftTask.java:101)
> > [23:22:33] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_18_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:483)
> > [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1547)
> > [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:480)
> > [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1470)
> > [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1264)
> > [23:22:33] [Server thread/WARN]: 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317)
> > [23:22:33] [Server thread/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
> > ```
> 
> wait... this looks like its using an old version of dough and not this branch.

this branch doesn't change dough-items/nms at all? it only updates dough-skins/nms

**Comment by md5sha256** [comment]
Ah, i misread the class name. Ok, ill have a look at fixing that.

**Review by TheBusyBiscuit — APPROVED** [review]

**Comment by TheBusyBiscuit** [comment]
@WalshyDev Re-review please. MockBukkit is also updated, we should be done after this.

**Review by WalshyDev — APPROVED** [review]
LGTM

**Pull request merged**